parts/k8s/cloud-init/artifacts/cse_config.sh

```diff
@@ -487,6 +487,33 @@ configAddons() {
 }
 {{end}}
 {{- if HasNSeriesSKU}}
+
+installGPUDriversRun() {
+    {{- /* there is no file under the module folder, the installation failed, so clean up the dirty directory
+    when you upgrade the GPU driver version, please help check whether the retry installation issue is gone,
+    if yes please help remove the clean up logic here too */}}
+    set -x
+    MODULE_NAME="nvidia"
+    NVIDIA_DKMS_DIR="/var/lib/dkms/${MODULE_NAME}/${GPU_DV}"
+    KERNEL_NAME=$(uname -r)
+    if [ -d "${NVIDIA_DKMS_DIR}" ]; then
```

Comment:
@andyliuliming @xuto2 Why does the `/var/lib/dkms/nvidia/418.40.04` directory exist sometimes but not all the time?

Reply:
it's created by the installer.

and if the first try of the installation failed, then the folder is in a dirty state.

and then when the second try of the installation, the installer will think the folder already there, and then fail fast. so we need to do some clean up.

Reply:
> and if the first try of the installation failed, then the folder is in a dirty state.

Where does the failure happen?

Reply:
from my observing, the failure happens if the nvidia installer choosed the "wrong" /lib/modules/kernel-name, if we have two kernel headers installed.

and then it will fail.
but it's not an always repro.

Reply:
and also it will happen if there's some trancient issues happened when the nvidia installer run.

and then it will keep on failing, the retry will not work if we do not do the clean up.

Reply:
> if the nvidia installer choosed the "wrong" /lib/modules/kernel-name

When does the above happen sometimes? What line of code?

Reply:
that's the "sh $GPU_DEST/nvidia-drivers-$GPU_DV --silent \
This conversation was marked as resolved by mboersma
        --kernel-name=${KERNEL_NAME} \
        --log-file-name=${log_file_name} \
        --accept-license --no-drm --dkms --utility-prefix="${GPU_DEST}" --opengl-prefix="${GPU_DEST}""


that's the installer command to install multiple things for nvidia.

the installer will choose the lib/modules/kernel-name, and then do the compilation and so on.

that command may fail due to either choosed the wrong lib-module, (there's some guessing logic in the installer), or due to some trancient issue.

and then it will have some dirty directory.

and when we retry to run the installer, then the installer will think there's already one module installed, but that's only a dirty directory, so the retry will fail again and again until reach the max retry times.

Reply:
and if we clean up the dirty directory, the retry will succeed.

Reply:
O.K., I see what's going on. This PR is making the execution of `/usr/local/nvidia/nvidia-drivers-418.40.04` retryable and idempotent. Is that correct?

Reply:
Thanks for the explanation. I see the `-k` option is only meant to be used when you're targeting a different kernel other than the one you're actually installing onto:

```
  -k KERNEL-NAME, --kernel-name=KERNEL-NAME
      Build and install the NVIDIA kernel module for the
      non-running kernel specified by KERNEL-NAME (KERNEL-NAME
      should be the output of `uname -r` when the target kernel
      is actually running).  This option implies
      '--no-precompiled-interface'.  If the options
      '--kernel-install-path' and '--kernel-source-path' are not
      given, then they will be inferred from KERNEL-NAME; eg:
      '/lib/modules/KERNEL-NAME/kernel/drivers/video/' and
      '/lib/modules/KERNEL-NAME/build/', respectively.
```

Does that mean that we don't trust the installer to choose the underlying runtime kernel that's actually on the VM?